### ISSUE TITLE
Adapting to large database schema

### ISSUE THREAD

**Issue by kotalaraghava** [ElementAI/picard#56]
Hey @tscholak, Thanks so much for this amazing work. It has been super helpful in my work.

I have been fiddling around Picard for the NL2SQL use case, I am able to successfully run with databases with 10 tables.

since we are encoding database scheme along with user query, I am wondering how will it scale in case 100 of tables and schema is big. 

- Does it become a limitation since language models can’t process tokens more than 512 (T5 transformers)?
- Does experimenting with the long former version of the T5 transformer will help in this case?
- Or is there any way of just passing input text along with chunks of the scheme or just passing text alone?

Any suggestions on how we can modify Picard for this use case?

**Comment by tscholak** [comment]
> Does it become a limitation since language models can’t process tokens more than 512 (T5 transformers)?

Yes, it will become a limitation, though differently than you (probably) think. T5 doesn't have a hard limited on the number of tokens it can take as input or produce as output. The model will be able to process more than 512 tokens, but with diminishing ability to align long-distance relationships in the input. If GPU memory is not a concern, then you should be able to go up to about 1000 tokens. Note, however, that there aren't enough big schemas in the Spider dataset on which the model could have learned to handle that much information. So, technically, large schemas are out of distribution, and the validation and test results are not going to tell you much about how well the model will work on that data.

> Does experimenting with the long former version of the T5 transformer will help in this case?

I don't know. S4 might, but afaik nobody has trained S4 on Spider and/or big database schemas.

> Or is there any way of just passing input text along with chunks of the scheme or just passing text alone?

There may be a clever pre-processing you can do to eliminate a lot of irrelevant tables and columns from the schema prior to encoding it, but such heuristics may not generalize well.

**Comment by tscholak** [comment]
Hey @kotalaraghava, unless there is something concrete to work on or fix, I'm inclined to close this issue. Feel free to reopen if necessary.
